File: services/client/src/protocol/batch.go
package protocol

import (
	"encoding/binary"
	"fmt"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
)

func SerializeBatch(bets []lottery.Bet, agencyId uint8) []byte {
	batchData := make([]byte, 0, MIN_BET_LEN)
	batchData = append(batchData, TYPE_BETS) // 1B
	batchData = append(batchData, agencyId)  // 1B

	for _, bet := range bets {
		batchData = append(batchData, SerializeBet(bet)...)
	}

	packet := make([]byte, 0, len(batchData)+2)
	packet = binary.BigEndian.AppendUint16(packet, uint16(len(batchData))) // 2B
	packet = append(packet, batchData...)

	return packet
}

func DeserializeBatch(batchData []byte) ([]lottery.Bet, error) {
	if len(batchData) < BATCH_HEADER_LEN {
		return nil, fmt.Errorf("packet too short")
	}

	messageType, err := ParseMessageType(batchData)
	if err != nil {
		return nil, err
	}

	if messageType != TYPE_BETS {
		return nil, fmt.Errorf("invalid bet type: %d", batchData[0])
	}

	agencyId := batchData[1]
	bets := []lottery.Bet{}

	for offset := BATCH_HEADER_LEN; offset < len(batchData); {
		bet, size, err := DeserializeBet(batchData[offset:], agencyId)
		if err != nil {
			return nil, err
		}
		bets = append(bets, bet)
		offset += size
	}
	return bets, nil
}

File: services/server/src/protocol/batch.py
from .bet import Bet, deserialize_bet, serialize_bet
from .message import TYPE_BETS, BATCH_HEADER_LEN, parse_message_type


def serialize_batch(bets: list[Bet], agency_id: int) -> bytes:
    bet_data = bytearray((TYPE_BETS, agency_id))
    for bet in bets:
        bet_data.extend(serialize_bet(bet))
    bet_data = len(bet_data).to_bytes(2, byteorder="big") + bet_data
    return bytes(bet_data)


def deserialize_batch(batch_data: bytes) -> list[Bet]:
    if len(batch_data) < BATCH_HEADER_LEN:
        raise ValueError("packet too short")

    message_type = parse_message_type(batch_data)
    if message_type != TYPE_BETS:
        raise ValueError(f"invalid bet type: {batch_data[0]}")

    agency_id = batch_data[1]
    bets = []
    offset = BATCH_HEADER_LEN

    while offset < len(batch_data):
        bet, size = deserialize_bet(batch_data[offset:], agency_id)
        bets.append(bet)
        offset += size
    return bets
